pep-9999.rst

```diff
@@ -0,0 +1,1102 @@
+semantic level. When a type parameter is shared among multiple generic
+classes, functions, and type aliases, all references are semantically
+equivalent.
+
+Type variables defined within the global scope also need to be given a name
+that starts with an underscore to indicate that the variable is private to
+the module. Globally-defined type variables are also often given names to
+indicate their variance leading to cumbersome names like "_T_contra" and
+"_KT_co". The current mechanisms for allocating type variables also requires
+the developer to supply a redundant name in quotes (e.g. ``T = TypeVar("T")``).
+This PEP eliminates the need for the redundant name and the cumbersome
+variable names.
+
+Defining type parameters today requires importing the ``TypeVar`` and 
+``Generic`` symbols from the ``typing`` module. Over the past several releases
+of Python, efforts have been made to eliminate the need to import ``typing``
+symbols for common use cases, and the PEP furthers this goal.
+
+
+Summary Examples
+================
+
+Defining a generic class prior to this PEP looks something like this.
+
+::
+
+    from typing import Generic, TypeVar
+
+    _T_co = TypeVar("_T_co", covariant=True, bound=str)
+
+    class ClassA(Generic[_T_co]):
+        ...
+
+
+With the new syntax, it looks like this.
+
+::
+
+    class ClassA[T: str]:
+        ...
+
+
+Here is an example of a generic function today.
+
+::
+
+    from typing import ParamSpec, TypeVar, Callable
+
+    _P = ParamSpec("_P")
+    _R = TypeVar("_R")
+
+    def func(cb: Callable[_P, _R], *args: _P.args, **kwargs: _P.kwargs) -> _R:
+        ...
+
+And the new syntax.
+
+::
+
+    from typing import Callable
+
+    def func[**P, R](cb: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
+        ...
+
+
+Here is an example of a generic type alias today.
+
+::
+
+    from typing import TypeAlias
+
+    _T = TypeVar("_T")
+
+    ListOrSet: TypeAlias = list[_T] | set[_T]
+
+
+And with the new syntax.
+
+::
+
+    type ListOrSet[T] = list[T] | set[T]
+
+
+Specification
+=============
+
+Type Parameter Declarations
+---------------------------
+
+We propose to add new syntax for declaring type parameters for generic
+classes, functions, and type aliases. The syntax adds support for
+a comma-delimited list of type parameters in square brackets after
+the name of the class, function, or type alias.
+
+Simple (non-variadic) type variables are declared with an unadorned name.
+Variadic type variables are preceded by ``*``. Parameter specifications are
+preceded by ``**``.
+
+::
+
+    # This generic class is parameterized by a TypeVar T, a
+    # TypeVarTuple Ts, and a ParamSpec P.
+    class ChildClass[T, *Ts, **P]: ...
+
+There is no need to include ``Generic`` as a base class. Its inclusion as
+a base class is implied by the presence of type parameters. The explicit
+use of a ``Generic`` base class will result in a runtime error.
```

Comment:
Will `typing.Generic` still appear in the `__mro__` of classes defined in this fashion? (It would be good to be explicit about that, either way.)

Reply:
Yes, it will still appear in the `__mro__` and `__orig_bases__`. I've updated the text to clarify this point.